4 subtests in `pointerevent_after_target_appended.html?mouse` are failing

lilyspiniolas opened the pull request:
#### dfc9d27f61ea947993d45d1d28ead144337fad3c
<pre>
4 subtests in `pointerevent_after_target_appended.html?mouse` are failing
<a href="https://bugs.webkit.org/show_bug.cgi?id=299560">https://bugs.webkit.org/show_bug.cgi?id=299560</a>
<a href="https://rdar.apple.com/161362257">rdar://161362257</a>

Reviewed by Wenson Hsieh and Abrar Rahman Protyasha.

The 4 subtests which test the events received after the child moves fail due to
a missing `pointerenter` or `mouseenter` event for the child. This is because,
before dispatching those events, we check if the target was an ancestor of the
previous element. However, the logic for this erroneously included the previous
element itself, resulting in the missing events. To resolve this, store
only the ancestors of the element under the mouse in m_ancestorsOfLastElementUnderMouse,
not the element itself.

Tested by
`imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_appended.html?mouse`.

* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_appended_interleaved.tentative_mouse-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_appended_interleaved.tentative_pen-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_appended_mouse-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_appended_pen-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_lostpointercapture_for_disconnected_node-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_pointer_boundary_events_after_reappending_last_over_target-expected.txt:
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::updateMouseEventTargetNode):
* Source/WebCore/page/EventHandler.h:

Canonical link: <a href="https://commits.webkit.org/300564@main">https://commits.webkit.org/300564@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bbf72224b22ecd62460dd13e2056063c9ecd82c1

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/123061 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/42775 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/33472 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/129717 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/75170 "Built successfully") | [❌ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/ffaf8060-114e-44af-9655-8e5fa49f4729) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/124938 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/43499 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/51370 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/93541 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/62080 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/47c199c4-5a7f-4f7a-bfcf-45bdc2f9b3fe) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/126012 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/34667 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/110137 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/74173 "Passed tests") | | [❌ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/a8f09301-2ad5-43a9-9806-aaebcc2a7b56) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/33642 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/28292 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/73231 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/104378 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/28518 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/132447 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/50011 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/38070 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/102041 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/50387 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/106358 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/101901 "Found 2 new API test failures: TestWTF:CompletionHandlerDeathTest.MainThreadHandlerOnThreadAssertsDeathTest, TestWTF:CompletionHandlerDeathTest.ConstructionThreadHandlerOnThreadAssertsDeathTest (failure)") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25897 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/47267 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/25468 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/46784 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/49866 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/55627 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/49334 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/52686 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/51015 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebCore/page/EventHandler.cpp:
```diff
@@ -3066,6 +3066,7 @@ void EventHandler::updateMouseEventTargetNode(const AtomString& eventType, Node*
             }
 
+            elementsUnderMouse.removeAt(0);
             m_ancestorsOfLastElementUnderMouse = WTFMove(elementsUnderMouse);
```
`m_ancestorsOfLastElementUnderMouse = elementsUnderMouse.subvector(1);` maybe?

lilyspiniolas replied:
I had to update it to make sure the vector is not empty before removing the first item, I think it might be slightly cleaner to do it in the current state now rather than using `subvector` unless there's a performance benefit?

aprotyas replied:
yeah, there isn't. Ignore!

lilyspiniolas committed:
4 subtests in `pointerevent_after_target_appended.html?mouse` are failing
https://bugs.webkit.org/show_bug.cgi?id=299560
rdar://161362257

Reviewed by Wenson Hsieh and Abrar Rahman Protyasha.

The 4 subtests which test the events received after the child moves fail due to
a missing `pointerenter` or `mouseenter` event for the child. This is because,
before dispatching those events, we check if the target was an ancestor of the
previous element. However, the logic for this erroneously included the previous
element itself, resulting in the missing events. To resolve this, store
only the ancestors of the element under the mouse in m_ancestorsOfLastElementUnderMouse,
not the element itself.

Tested by
`imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_appended.html?mouse`.

* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_appended_interleaved.tentative_mouse-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_appended_interleaved.tentative_pen-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_appended_mouse-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_appended_pen-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_lostpointercapture_for_disconnected_node-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_pointer_boundary_events_after_reappending_last_over_target-expected.txt:
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::updateMouseEventTargetNode):
* Source/WebCore/page/EventHandler.h:

Canonical link: https://commits.webkit.org/300564@main
